File: zhangyansong/4sum.rb
# Given an array S of n integers, are there elements a, b, c, and d in S such that a + b + c + d = target? Find all unique quadruplets in the array which gives the sum of target.

# Note: The solution set must not contain duplicate quadruplets.

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[][]}
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[][]}
def four_sum(nums, target)
    return [] if nums.size() < 4
    nums.sort!
    num_last = nil
    result = Array.new
    max = nums[-1] * 3
    nums.each_with_index do |num, idx|
        next if num == num_last
        next if num + max < target # too small
        break if num * 4 > target # too large
        target_result = target - num
        three_sum(nums, idx+1, target_result, result)
        num_last = num
    end
    result
end

def three_sum(nums, idx_from_four, target_result, result)
    num_last = nil
    max = nums[-1] * 2
    (idx_from_four..nums.size()-2).each do |idx|
        next if nums[idx] == num_last
        next if nums[idx] + max < target_result # too small
        break if nums[idx] * 3 > target_result # too large
        target = target_result - nums[idx]
        two_sum(nums, idx_from_four, idx+1, target, result)
        num_last = nums[idx]
    end
end

def two_sum(nums, idx_from_four, start, target, result)
  head, tail = start, nums.size() - 1
  while head < tail
    tmp = nums[head] + nums[tail]
    break if nums[tail] * 2 < target # too small 
    break if nums[head] * 2 > target # too large
    if tmp < target
      head += 1
    elsif tmp > target
      tail -= 1
    else
      result << [nums[idx_from_four-1], nums[start-1], nums[head], nums[tail]]
      tmp = head + 1
      while tmp < tail && nums[tmp] == nums[head]
        tmp += 1
      end
      head = tmp
      tmp = tail - 1
      while tmp > head && nums[tmp] == nums[tail]
        tmp -= 1
      end
      tail = tmp
    end
  end
end

File: caoguodong/4sum.java
public class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> result = new ArrayList<>();

        if(nums.length < 4){
            return result;
        }

        Arrays.sort(nums);
        
        int b,c,d;
        for(int i=0; i<nums.length-3;i++){
            b=i+1;
            c=b+1;
            d=c+1;
            
            if(i == 0 || nums[b]>nums[i]){
                while(b<nums.length-3){
                        if(nums[i]+nums[b]+nums[c]+nums[d]==target){
                            if(d== 3 || nums[c]<nums[d]){
                        List<Integer> list = new ArrayList<>();
                        list.add(nums[i]);
                        list.add(nums[b]);
                        list.add(nums[c]);
                        list.add(nums[d]);
                        result.add(list);         
                            }
                         
                        
                        
                        b++;
                        c=b+1;
                        d=c+1;
                        } else if(nums[i]+nums[b]+nums[c]+nums[d]<target){
                            if(d< nums.length-1){
                                d++;                    
                            } else if (d== nums.length -1 && c<nums.length-2){
                                c++;
                                d=c+1;
                            } else if(d== nums.length -1 && c==nums.length-2){
                                b++;
                                c=b+1;
                                d=c+1;
                            }
                        } else{
                            break;
                        }
                    }
            }
        }
        return result;
    }
}